nightfall-optimist/src/services/transaction-sign-send.mjs

```diff
@@ -0,0 +1,84 @@
+/* eslint-disable no-param-reassign */
+
+import config from 'config';
+import { web3 } from '@polygon-nightfall/common-files/utils/contract.mjs';
+import { estimateGas, estimateGasPrice } from '@polygon-nightfall/common-files/utils/gas.mjs';
+import logger from '@polygon-nightfall/common-files/utils/logger.mjs';
+import { Mutex } from 'async-mutex';
+
+const { GAS, GAS_MULTIPLIER, GAS_ESTIMATE_ENDPOINT, GAS_PRICE, GAS_PRICE_MULTIPLIER } = config;
+
+const nonceMutex = new Mutex();
+
+const isListening = async () => {
+  try {
+    await web3.eth.net.isListening();
+  } catch (error) {
+    throw new Error('Web3 ws is disconnected, try again later');
+  }
+};
+
+export async function getAddressNonce(ethAddress) {
+  let nonce;
+  try {
+    nonce = await web3.eth.getTransactionCount(ethAddress);
+  } catch (err) {
+    logger.error({
+      msg: 'Error obtaining address nonce',
+      err,
+    });
+  }
+  return nonce;
+}
+
+// TODO document
+export async function createSignedTransaction(nonce, ethPrivateKey, from, to, data, value = 0) {
+  // Check if web3 ws is available
+  await isListening();
+
+  logger.debug({ msg: 'Create transaction object...' });
+
+  let signedTx;
+  await nonceMutex.runExclusive(async () => {
+    // Update nonce if necessary
+    const _nonce = await getAddressNonce(from);
+    if (nonce < _nonce) {
```

Comment:
why is this needed?

Reply:
I think this is to synchronize the nonce if another transaction has been sent before processing this transaction. If not then a revert will be thrown because the account has a higher nonce.

Reply:
@daveroga isn't that what nonceMutex.runExclusive() is taking care of?